airflow/contrib/operators/gcp_vision_operator.py

```diff
@@ -671,3 +672,284 @@ def execute(self, context):
+    # [END vision_annotate_image_template_fields]
+
+    @apply_defaults
+    def __init__(
+        self, request, retry=None, timeout=None, gcp_conn_id='google_cloud_default', *args, **kwargs
+    ):
+        super(CloudVisionAnnotateImageOperator, self).__init__(*args, **kwargs)
+        self.request = request
+        self.retry = retry
+        self.timeout = timeout
+        self.gcp_conn_id = gcp_conn_id
+
+    def execute(self, context):
+        hook = CloudVisionHook(gcp_conn_id=self.gcp_conn_id)
+        return hook.annotate_image(request=self.request, retry=self.retry, timeout=self.timeout)
+
+
+class CloudVisionReferenceImageCreateOperator(BaseOperator):
+    """
+    Creates and returns a new ReferenceImage ID resource.
+
+    .. seealso::
+        For more information on how to use this operator, take a look at the guide:
+        :ref:`howto/operator:CloudVisionReferenceImageCreateOperator`
+
+    :param location: (Required) The region where the Product is located. Valid regions (as of 2019-02-05) are:
+        us-east1, us-west1, europe-west1, asia-east1
+    :type location: str
+    :param reference_image: (Required) The reference image to create. If an image ID is specified, it is
+        ignored.
+        If a dict is provided, it must be of the same form as the protobuf message
+        :class:`google.cloud.vision_v1.types.ReferenceImage`
+    :type reference_image: dict or google.cloud.vision_v1.types.ReferenceImage
+    :param reference_image_id: (Optional) A user-supplied resource id for the ReferenceImage to be added.
+        If set, the server will attempt to use this value as the resource id. If it is already in use, an
+        error is returned with code ALREADY_EXISTS. Must be at most 128 characters long. It cannot contain
+        the character `/`.
+    :type reference_image_id: str
+    :param product_id: (Optional) The resource id of this Product.
+    :type product_id: str
+    :param project_id: (Optional) The project in which the Product is located. If set to None or
+        missing, the default project_id from the GCP connection is used.
+    :type project_id: str
+    :param retry: (Optional) A retry object used to retry requests. If `None` is
+        specified, requests will not be retried.
+    :type retry: google.api_core.retry.Retry
+    :param timeout: (Optional) The amount of time, in seconds, to wait for the request to
+        complete. Note that if retry is specified, the timeout applies to each individual
+        attempt.
+    :type timeout: float
+    :param metadata: (Optional) Additional metadata that is provided to the method.
+    :type metadata: sequence[tuple[str, str]]
+    :param gcp_conn_id: (Optional) The connection ID used to connect to Google Cloud Platform.
+    :type gcp_conn_id: str
+    """
+
+    # [START vision_reference_image_create_template_fields]
+    template_fields = ('location', 'product_id', 'reference_image_id', 'project_id', 'gcp_conn_id')
```

Comment:
There may be a need to be able to render reference_image's URL using template, then we can use "task_instance.xcom_pull(...)". It would be useful if this reference image is created by a precedent task.

For example:
reference_image = ReferenceImage(uri=GCP_VISION_REFERENCE_IMAGE_URL)
The URL "GCP_VISION_REFERENCE_IMAGE_URL" may be given by a precedent task.

There may be some complexity when for reference_image that is a "google.cloud.vision_v1.types.ReferenceImage" instead of a dict.

It is similar for CloudVisionAnnotateImageOperator that the input image URL may be generated by another task, and it already supports image URL template there.

Reply:
Good point. It makes perfect sense. I think this is not a problem for the Protobuf types. We should simply mention in the doc that templates for reference_image can only be used if you pass reference image as dict. I think that is good enough.

Reply:
> Good point. It makes perfect sense. I think this is not a problem for the Protobuf types. We should simply mention in the doc that templates for reference_image can only be used if you pass reference image as dict. I think that is good enough.

I just found that it is no problem for the Protobuf types as well. Actually any serializable types work here. Sorry for the confusion.

Reply:
I think this is still a problem and should be mentioned in the docs. The code to render templates is here: https://github.com/apache/airflow/blob/master/airflow/models/__init__.py#L2447  and it seems to support only strings and nested lists/tuples/dicts of strings. I checked that for protobuf-generated objects it will return it "as is" without rendering the templates.

Reply:
This can change:
https://github.com/apache/airflow/pull/4743/files

Reply:
I added reference_image to template_fields.